### PR TITLE
KT-30975 ''when' has only 'else' branch and should be simplified' inspection removes subject variable definition used in else branch

### PR DESCRIPTION
**Pull request by t-kameyama**
https://youtrack.jetbrains.com/issue/KT-30975

### CODE REVIEW

File: idea/src/org/jetbrains/kotlin/idea/inspections/ConstantConditionIfInspection.kt [hunk 1]
```diff
@@ -157,27 +159,59 @@ private fun KtExpression.enumEntry(): KtEnumEntry? {
...
 fun KtExpression.replaceWithBranch(branch: KtExpression, isUsedAsExpression: Boolean, keepBraces: Boolean = false) {
-    val lastExpression = when {
-        branch !is KtBlockExpression -> replaced(branch)
+    val caretModel = findExistingEditor()?.caretModel
+
+    val subjectVariable = (this as? KtWhenExpression)?.subjectVariable?.let { it ->
+        if (it.annotationEntries.isNotEmpty()) return@let it
+        val initializer = it.initializer ?: return@let it
+        val references = ReferencesSearch.search(it, LocalSearchScope(this)).toList()
+        if (references.size > 1) return@let it
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ConstantConditionIfInspection.kt]: I think it'd be more useful to replace `toList` with `singleOrNull`.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ConstantConditionIfInspection.kt [hunk 2]
```diff
@@ -157,27 +159,59 @@ private fun KtExpression.enumEntry(): KtEnumEntry? {
...
+    }
+
+    val factory = KtPsiFactory(this)
+    val parent = this.parent
+    val replaced = when {
+        branch !is KtBlockExpression -> {
+            if (subjectVariable != null) {
+                if (isUsedAsExpression) {
+                    replaced(KtPsiFactory(this).createExpressionByPattern("run { $0\n$1 }", subjectVariable, branch))
+                } else {
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ConstantConditionIfInspection.kt]: I think we shouldn't expand the scope for subjectVariable.
For example:
``` Kotlin
when (val a = 42) {
    else -> use(a, a)
}

val a = 33
```

Comment (t-kameyama) [reply idea/src/org/jetbrains/kotlin/idea/inspections/ConstantConditionIfInspection.kt]: I want to return null if references size is 0.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ConstantConditionIfInspection.kt (hunk 1 above)
Comment (t-kameyama) [reply idea/src/org/jetbrains/kotlin/idea/inspections/ConstantConditionIfInspection.kt]: I want to return null if references size is 0.

File: idea/src/org/jetbrains/kotlin/idea/inspections/WhenWithOnlyElseInspection.kt [hunk 3]
```diff
@@ -24,15 +30,28 @@ class WhenWithOnlyElseInspection : AbstractKotlinInspection() {
 
             val usedAsExpression = expression.isUsedAsExpression(expression.analyze())
 
-            holder.registerProblem(expression,
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/WhenWithOnlyElseInspection.kt]: I think we shouldn't change this file.

File: idea/testData/inspectionsLocal/whenWithOnlyElse/subjectVariable/blockElse/noReference.kt.after [hunk 4]
```diff
@@ -0,0 +1,10 @@
+fun test() {
+    <caret>use("")
```
Comment (dimonchik0036) [inline idea/testData/inspectionsLocal/whenWithOnlyElse/subjectVariable/blockElse/noReference.kt.after]: We can remove `create()` only if it hasn't side effects.

File: idea/testData/inspectionsLocal/whenWithOnlyElse/subjectVariable/hasSameNameVariable.kt [hunk 5]
```diff
@@ -0,0 +1,9 @@
+// PROBLEM: none
+fun test() {
```
Comment (dimonchik0036) [inline idea/testData/inspectionsLocal/whenWithOnlyElse/subjectVariable/hasSameNameVariable.kt]: We should convert to
```kotlin
fun test() {
    run {
        val a = 42
        use(a, a)
    }
    val a = 33
}
```

**Comment by dimonchik0036** [comment]
We do not have good tools to check for side effects now, so this is another issue. I accept this PR.

**Comment by dimonchik0036** [comment]
Thanks! e955dcfc147bffc9121610cf5e9a594f789ba520

**Pull request closed without merging**